src/communication/websocket/session.cpp

```diff
@@ -88,13 +91,40 @@ void Session::DoRead() {
       }));
 }
 
-void Session::OnRead(boost::beast::error_code ec, size_t /*bytes_transferred*/) {
+bool Session::Authenticate() {
+  try {
+    const auto creds = nlohmann::json::parse(boost::beast::buffers_to_string(buffer_.data()));
+    if (auth_.Authenticate(creds.at("username").get<std::string>(), creds.at("password").get<std::string>())) {
+      authenticated_ = true;
+      return true;
+    }
+  } catch (const nlohmann::json::out_of_range &out_of_range) {
+    spdlog::error("Missing username or password: {}!", out_of_range.what());
+  } catch (const nlohmann::json::parse_error &parse_error) {
+    spdlog::error("Cannot parse JSON containing username and password: {}!", parse_error.what());
+  }
+  return false;
+}
+
+void Session::OnRead(const boost::beast::error_code ec, const size_t /*bytes_transferred*/) {
   if (ec == boost::beast::websocket::error::closed) {
     messages_.clear();
     connected_.store(false, std::memory_order_relaxed);
     return;
   }
 
+  if (!authenticated_) {
+    auto response = nlohmann::json();
+    if (Authenticate()) {
+      response["success"] = true;
+      response["message"] = "User has been successfully authenticated!";
+    } else {
+      response["success"] = false;
+      response["message"] = "Authentication failed!";
+    }
+    Write(std::make_shared<std::string>(response.dump()));
```

Comment:
I would close the connection here.
I see the authentication JSON as part of the handshake. Even though the WS passed, we still don't know if the user is allowed to interact with us using that connection in any way.
But I cannot say anything for sure, maybe it's just an extra inconvenience for the client-side without extra benefits.

@tonilastre What do you think?
Lab <-> Memgraph make a WebSocket connection. 
Lab sends a JSON with auth details as the first message, but it's invalid. 
Would you destroy the WS connection or keep it alive so Memgraph gives Lab multiple chances to send valid credentials on the same established connection?

Reply:
Hmm. Both ways would work. The Lab will send the credentials only once because we will establish a WS connection after successfully connecting via bolt (our standard connection).
From that point, closing a connection or keeping it alive will be the same because we won't try it out again. The user needs to log out in the Lab and try out new credentials which will definitely yield a new WS connection.

Even though the purpose of the WS connection is the reuse it as much as you can, this is somehow a special case because of the handshake.

**Future-proof**
In the future, I am seeing Lab as purely a web product that won't use bolt but WS/WSS protocol and everything will go through it - with this in mind, it still makes more sense to kill the connection server-side after incorrect credentials. Here is why: Let's say I have a public Memgraph, then someone can just create WS connections towards my Memgraph and send incorrect credentials. I will soon go out of memory because every WS connection (unauthorized) will be keep-alive. We don't want that.

**Recap**
I would suggest destroying the WS connection after the invalid credentials.

Reply:
Oh, the `Future-proof` point makes a lot of sense. We would like to give users the ability to connect to that WS server by themselves, not just through Lab. So a scenario you described can easily happen.

@jbajic We already agreed on that but just to have it here, we should try destroying the connection if the authorization fails.